src/reducer.js

```diff
@@ -13,38 +13,71 @@ function startHere(state: {} = {}) {
   return state;
 }
 
+function addTodo(
+  state: Array<{
+    uid: string,
+    target: {},
+    actionId: string,
+    completed: boolean}>,
+  actionId: string,
+  phoneNumber: {}) {
+  // If there are already todos for this action, do nothing
+  if (_.find(state, { actionId, target: phoneNumber })) {
+    return state;
+  }
+  return state.concat(
+    {
+      actionId,
+      target: phoneNumber,
+      uid: uuid.v4(),
+      completed: false,
+    },
+  );
+}
+
+function toggleTodo(
+  state: Array<{
+    uid: string,
+    target: {},
+    actionId: string,
+    completed: boolean}> = [],
+  actionId: string,
+  phoneNumber: {}) {
+  return state.map((todo) => {
+    if (todo.actionId === actionId && todo.target === phoneNumber) {
+      return Object.assign({}, todo, {
+        completed: !todo.completed,
+      });
+    }
+    return todo;
+  });
+}
+
 function todos(
   state: Array<{
     uid: string,
     target: {},
     actionId: string,
     completed: boolean}> = [],
   action: {
-    type: string, actionId?: string, uid?: string, phoneNumbers?: Array<{}>,
+    type: string,
+    actionId: string, uid?: string,
+    phoneNumbers?: Array<{}>,
+    phoneNumber?: {},
   }) {
   switch (action.type) {
+    case ADD_OR_TOGGLE_TODO:
+      return toggleTodo(addTodo(state, action.actionId, action.phoneNumber),
```

Comment:
If you add a [null check](https://flowtype.org/docs/nullable-types.html), flow will infer that `phoneNumber` now has to be an object. Unfortunately, the code for this is a bit awkward because flow is [uneasy about null checks on object properties](https://flowtype.org/docs/nullable-types.html#maybe-and-objects). This is what I would do:

```js
  const phoneNumber = action.phoneNumber;
  switch (action.type) {
    case ADD_OR_TOGGLE_TODO:
      if (phoneNumber == null) throw new Error("Can't add / toggle a todo without a phone number");
      return toggleTodo(addTodo(state, action.actionId, phoneNumber), action.actionId, phoneNumber);

  ...
```